static/js/passed-state.js

```diff
@@ -0,0 +1,131 @@
+class PassedState {
+    _key = 'python-type-challenges';
+
+    /**
+     * Initializing when there is no state in the local storage. If there is no state in the local storage, the initial state is required.
+     * this function will check the new state and the old state whether is undefined or not and updated the old state based on new state.
+     * 
+     * @param {object} initialState - the initial state of the challenges which grouped by the level.
+     * @returns void
+     */
+    init(initialState) {
+        const currentState = this.get();
+        // initialize the state when there is no state in the local storage.
+        if (!currentState && !initialState) {
+            throw new Error('initial state is required when there is no state in the local storage.');
+        }
+
+        // add passed property to the challenges in the initial state.
+        const rawState = this._prepareState(initialState);
+
+        // check new state and old state whether is undefined or not. and merge the new state to the old state.
+        const state = this._checkAndMerge(currentState, rawState);
+        this._save(state);
+    }
+    /**
+     * prepare the state for initialization.
+     * 
+     * @param {object} rawState 
+     * @returns state - the state contains the challenge name and whether the challenge is passed.
+     */
+    _prepareState(rawState) {
```

Comment:
Since there's only one callsite for this function, maybe it's better to fold it into `_checkAndMerge`, to make it more readable.